pkg/mcclient/controller.go

```diff
@@ -0,0 +1,179 @@
+	logger                 logr.Logger
+	queue                  workqueue.RateLimitingInterface
+	logicalClusterInformer cache.SharedIndexInformer
+	multiClusterClient     *multiClusterClient
+}
+
+func newController(
+	context context.Context,
+	logicalClusterInformer cache.SharedIndexInformer,
+	multiClusterClient *multiClusterClient,
+) *controller {
+	context = klog.NewContext(context, klog.FromContext(context).WithValues("controller", ControllerName))
+
+	c := &controller{
+		context:                context,
+		logger:                 klog.FromContext(context),
+		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+		logicalClusterInformer: logicalClusterInformer,
+		multiClusterClient:     multiClusterClient,
+	}
+
+	logicalClusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueue,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldInfo := oldObj.(*v1alpha1.LogicalCluster)
+			newInfo := newObj.(*v1alpha1.LogicalCluster)
+			if !reflect.DeepEqual(oldInfo.Status, newInfo.Status) {
+				c.enqueue(newObj)
+			}
+		},
+		DeleteFunc: c.enqueue,
+	})
+
+	return c
+}
+
+func (c *controller) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+// Run starts the controller, which stops when c.context.Done() is closed.
+func (c *controller) Run(numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("starting client logicalcluster controller")
+	defer c.logger.Info("shutting down client logicalcluster controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(c.context, c.runWorker, time.Second)
+	}
+
+	<-c.context.Done()
+}
+
+func (c *controller) runWorker(ctx context.Context) {
+	for c.processNextItem() {
+	}
+}
+
+func (c *controller) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	keyRaw := key.(string)
+
+	// Done with this key, unblock other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(keyRaw); err != nil {
+		runtime.HandleError(err)
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(key string) error {
+	cluster, exists, err := c.logicalClusterInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		c.handleDelete(key)
+	} else {
+		c.handleAdd(cluster)
+	}
+	return nil
+}
+
+func (c *controller) handleAdd(cluster interface{}) {
+	clusterInfo, ok := cluster.(*v1alpha1.LogicalCluster)
+	if !ok {
+		runtime.HandleError(errors.New("unexpected object type. expected LogicalCluster"))
+		return
+	}
+	// only clusters in ready state are cached
+	if clusterInfo.Status.Phase != v1alpha1.LogicalClusterPhaseReady {
+		c.handleDelete(clusterInfo.Name)
```

Comment:
Why delete ?  Don't we need to just wait till Ready, e.g., ignore this and wait for additional status change

Reply:
It only deletes from the cache maps. If `Status.Phase` is not ready then the cluster is not accessible and config/cs is not valid. Client code which try to use this config/cs will fail. We will get another event when the cluster is Ready and then we cache it.

Reply:
Thanks. I forgot you use the same function for Update as well. Can you please add a comment (copy from your reply :-) ) ?

Reply:
Done.